enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,340 @@
+
+### API Extensions
+
+This is adding new Insights CRD in the `insights.openshift.io/v1alpha1` . The proposal of the structure of the new CRD is mentioned in the [Proposal](#Proposal) section.
+
+### Implementation Details/Notes/Constraints
+
+#### Current behavior
+The Insights operator currently runs and gathers the data in a single container. 
+Uploading of new archives, pruning old archives as well as downloading the last Insights analysis report (this is always per archive) is implemented in this one container/binary. 
+When there is a successful upload (to the console.redhat.com) of the archive then there is some short delay (waiting for the processing in console.redhat.com pipelines - see the [API documentation](https://console.redhat.com/docs/api/insights-results-aggregator/v2)) and then there is a request to get the last Insights analysis report per given cluster. 
+Finally the status section of the `insightsoperator.operator.openshift.io` CR is updated. Note that there is a retry logic in case of failed upload and after some time (and number of attempts) the Insights operator is marked as degraded.
+
+#### Updates required to allow on-demand data gathering
+The plan is to refactor the existing periodic data gathering to run in a separate Pod/Job. The existing Insights operator pod will still exist, but it would periodically do the following:
+- check the `insightsdatagather.config.openshift.io` CR to read current user configuration
+- create a new `datagather.insights.openshift.io` CR (in the `openshift-insights` namespace) based on the values from the above
+- spawn a new `job` to gather the Insights data, upload it to `console.redhat.com` and check that the archive was processed in console.redhat.com (this requires a new API endpoint to be implemented in https://console.redhat.com/docs/api/insights-results-aggregator/v2).
+  The new API endpoint will be identified by the cluster ID and will require `insightsRequestID` parameter. The endpoint will be queried with some delay (to allow processing)  
+- status attribute of the `datagather.insights.openshift.io` is updated within the job execution
+- wait for the job completion and if the processing part was successful then download the last Insights analysis report
+- check that the Insights analysis report ID is equal to the gathering job request ID. If the IDs are equal then update the corresponding job status with `AnalysisAvailable=True` condition and continue with the following step below, otherwise update the corresponding job status with `AnalysisAvailable=False` condition and do nothing 
+- based on the status of the last `datagather.insights.openshift.io` CR update the `insightsoperator.operator.openshift.io` CR and also the clusteroperator conditions
+- prune the old job definitions and `datagather.insights.openshift.io` CRs
```

Comment:
how aggressive is this pruning?

Reply:
I think we agreed on 24 hours? Is that not documented somewhere in here? (edit line 271, open questions answers this)

Reply:
Yes. We have 24h, but I was thinking about doing it even more aggresive by default. Let's say 12h. WDYT?

Reply:
IMO 24 hours should be fine, IIRC that's what CSR pruning is so it would tie in with that well. Realistically we are expecting approx 12 resources on any random day? Plus a few more if we are triggering some on demand? I think sub 30 resources is still a reasonable amount to keep around, but open to other's suggestions

Reply:
Yes 12 jobs + 12 CRs on random day plus few more if someone triggered on demand gathering.

Reply:
Unless David has a strong opinion, my suggestion is to stick to 24h